src/components/FallenHeroes.tsx

```diff
@@ -20,23 +34,27 @@ function FallenHeroes() {
   function addFallenHero(e: React.FormEvent<HTMLFormElement>) {
     e.preventDefault();
     //add hero to the list if it isnt there yet
-    if (!campaign.restOfAliveHeroes.includes(fallenHero)) {
+    if (!names.includes(fallenHero)) {
+      throw Error("Sorry, your hero is an unknown!");
+    }
+
+    if (names.includes(fallenHero) && !isAlive) {
       throw Error("Sorry, your hero is not among the living!");
     }
-    if (campaign.restOfAliveHeroes.includes(fallenHero)) {
-      const updatedFallen = [...campaign.fallenHeroes, fallenHero];
-      const updatedRestOfAlive = campaign.restOfAliveHeroes.filter(
-        (hero) => hero !== fallenHero
-      );
-      setCampaign({
-        ...campaign,
-        fallenHeroes: updatedFallen,
-        restOfAliveHeroes: updatedRestOfAlive,
+
+    if (names.includes(fallenHero) && isAlive) {
+      campaign.allHeroes.map(function (hero) {
+        if (hero.name === fallenHero) {
+          hero.alive = false;
+          setCampaign(campaign);
+          const fallen = filterFallen();
+          setFallenHeroes(fallen);
+        }
+        return campaign;
```

Comment:
There is no point in returning a value here as you do not assign it to anything. `Array.prototype.map` is meant for creating a modified array, which is not the goal here, you only want to change one item.  You can `find` the item and modify it. It be more readable and also more efficient as you won't have to loop through the whole array.